core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java

```diff
@@ -129,31 +182,188 @@ public FieldSortBuilder setNestedPath(String nestedPath) {
+                Objects.equals(this.sortMode, builder.sortMode) &&
+                Objects.equals(this.unmappedType, builder.unmappedType));
+    }
+    
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.fieldName, this.nestedFilter, this.nestedPath,
+                this.missing, this.order, this.sortMode, this.unmappedType);
+    };
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(this.fieldName);
+        if (this.nestedFilter != null) {
+            out.writeBoolean(true);
+            out.writeQuery(this.nestedFilter);
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeOptionalString(this.nestedPath);
+        if (this.missing != null) {
+            out.writeBoolean(true);
+            out.writeGenericValue(this.missing);
+        } else {
+            out.writeBoolean(false);
+        }
+        
+        if (this.order != null) {
+            out.writeBoolean(true);
+            this.order.writeTo(out);
+        } else {
+            out.writeBoolean(false);
+        }
+        
+        out.writeOptionalString(this.sortMode);
+        out.writeOptionalString(this.unmappedType);
+    }
+
+    @Override
+    public FieldSortBuilder readFrom(StreamInput in) throws IOException {
+        String fieldName = in.readString();
+        FieldSortBuilder result = new FieldSortBuilder(fieldName);
+        if (in.readBoolean()) {
+            QueryBuilder query = in.readQuery();
+            result.setNestedFilter(query);
+        }
+        result.setNestedPath(in.readOptionalString());
+        if (in.readBoolean()) {
+            result.missing(in.readGenericValue());
+        }
+        if (in.readBoolean()) {
+            result.order(SortOrder.readOrderFrom(in));
+        }
+        result.sortMode(in.readOptionalString());
+        result.unmappedType(in.readOptionalString());
+        return result;
+    }
+
+    @Override
+    public FieldSortBuilder fromXContent(QueryParseContext context) throws IOException {
+        XContentParser parser = context.parser();
+
+        String fieldName = null;
+        QueryBuilder nestedFilter = null;
+        String nestedPath = null;
+        Object missing = null;
+        SortOrder order = null;
+        String sortMode = null;
+        String unmappedType = null;
+        
+        String currentFieldName = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+                if (fieldName != null) {
+                    throw new ParsingException(parser.getTokenLocation(),
+                            "[term] query does not support different field names, use [bool] query instead");
```

Comment:
This seems not to be the right exception message here (looks like cp'ed from term query).